### PR TITLE
Codefresh sync

### PR DESCRIPTION
**Pull request by goruha**
## What
* Added targets to sync codefresh pipelines from code with templates

## Why
* To have a repeatable, stable  and scalable process of creation/update codefresh pipelines

## Usage
* Create new repo that would contain templates for your pipelines
```
{repo}                                                           # Repository for codefresh sync
├── templates                                            # Directory for templates
│   ├── {account_x}                                    # Templates for account x
│   │    ├── {template_x_x1}.yaml              # Pipeline template {template_x_x1}
│   │    └── {template_x_y1}.yaml                # Pipeline template {template_x_y1}
│   └── {account_y}                                      # Templates for account y
│        ├── {template_y_x1}.yaml                # Pipeline template {template_y_x1}
│        └── {template_y_y1}.yaml                # Pipeline template {template_y_y1}
└── Makefile                                               # Makefile
```
* Fill templates with [pipelines](https://codefresh-io.github.io/cli/pipelines/spec/). You can use `{{ (ds "repository") }}` as placeholder for repo name
* Add to the `Makefile` 
```
export DEFAULT_HELP_TARGET ?= help
export HELP_FILTER ?= help|init|codefresh/sync
export DOCKER_NETWORK ?= default
export BUILD_HARNESS_BRANCH ?= 0.14.5
-include $(shell curl -sSL -o .build-harness "https://git.io/build-harness"; echo .build-harness)

## Project name
PROJECT ?= project_z

## Repos to sync pipelines
REPOSITORIES ?= api \
				front
## Codefresh accounts to operate with
ACCOUNTS ?= account_x account_y

## Pipelines for `acccount_x`
ACCOUNT_X_PIPELINES ?= template_x_x1 template_x_y1

## Pipelines for `acccount_y`
ACCOUNT_Y_PIPELINES ?= template_y_x1 template_y_y1
```
* Do Auth for accounts. You can get `CF_API_KEY` [here](https://g.codefresh.io/account-conf/tokens)
```
make codefresh/sync/auth/account_x CF_API_KEY={CF_API_KEY for `account_x`}
make codefresh/sync/auth/account_y CF_API_KEY={CF_API_KEY for `account_y`}
```

* For dry-run use
```
make codefresh/sync
```

* For sync use 
```
make codefresh/sync APPLY=true
```

* You can filter repos, accounts and pipelines that are syncing with env variables `REPOSISTORIES` `ACCOUNTS` `PIPELINES`

```
make codefresh/sync REPOSITORIES=api ACCOUNTS=account_x PIPELINES=template_x_x1
```

### CODE REVIEW

File: bin/codefresh.sync.sh [hunk 1]
```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env bash
+
+## Messages templates
+MESSAGE_SEPARATOR="_______________________________________________________________________________________________";
```
Comment (osterman) [inline bin/codefresh.sync.sh]:
```suggestion
MESSAGE_SEPARATOR="$(printf -- '_%.0s' {1..80})";
```

File: bin/codefresh.sync.sh [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	## Compare masked pipelines to be indifferent to timestamps and ids
+	yq m -x -i ${PIPELINE_CURRENT} ${PIPELINE_MACK}
+	yq m -x -i ${PIPELINE_NEW} ${PIPELINE_MACK}
+fi
+
+## Get diff between current and new pipelines
+CODEFRESH_SYNC_PIPELINE_DIFF=$(diff -u -s --suppress-common-lines ${PIPELINE_CURRENT} ${PIPELINE_NEW})
+CODEFRESH_SYNC_PIPELINE_DIFF_IS_EMPTY=$?
+
+if [[ "${CODEFRESH_SYNC_PIPELINE_DIFF_IS_EMPTY}" == "0" ]]
```
Comment (osterman) [inline bin/codefresh.sync.sh]:
```suggestion
if [[ "${CODEFRESH_SYNC_PIPELINE_DIFF_IS_EMPTY}" == "0" ]]; then
```

https://google.github.io/styleguide/shell.xml?showone=Loops#Loops

File: bin/codefresh.sync.sh [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+then
+	echo "${MESSAGE_NO_CHANGES_REQUIRED}"
+else
+	if [[ "${APPLY}" == "true" ]]
```
Comment (osterman) [inline bin/codefresh.sync.sh]:
```suggestion
	if [[ "${APPLY}" == "true" ]]; then
```
https://google.github.io/styleguide/shell.xml?showone=Loops#Loops

File: bin/codefresh.sync.sh [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	then
+
+		if [[ "${PIPELINE_IS_NEW}" == "1" ]]
```
Comment (osterman) [inline bin/codefresh.sync.sh]:
```suggestion
		if [[ "${PIPELINE_IS_NEW}" == "1" ]]; then
```

https://google.github.io/styleguide/shell.xml?showone=Loops#Loops

File: modules/codefresh/Makefile.sync [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+		done; \
+		echo "============================================================================================================" ; \
+	done
+
+codefresh/sync/apply:
+	$(call assert-set,REPOSITORY)
+	$(call assert-set,ENVIRONMENT)
+	$(call codefresh-use-context,$(ENVIRONMENT))
+	@for pipeline in $(PIPELINES); do \
+		PIPELINE=$$pipeline $(BUILD_HARNESS_PATH)/bin/codefresh.sync.sh; \
```
Comment (osterman) [inline modules/codefresh/Makefile.sync]:
```suggestion
		PIPELINE=$$pipeline $(BUILD_HARNESS_PATH)/bin/codefresh-sync.sh; \
```

File: templates/codefresh.sync.mask.yaml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+metadata:
+  accountId: xxx
+  created_at: xxx
```
Comment (osterman) [inline templates/codefresh.sync.mask.yaml:3]: they mix and match camel case with snake case?

File: modules/codefresh/Makefile.sync [hunk 7]
```diff
@@ -0,0 +1,46 @@
+CODEFRESH_CLI_IMAGE ?= codefresh/cli
+CODEFRESH_CLI_VERSION ?= latest
+
+
+export CODEFRESH_CLI ?= $(DOCKER_RUN) -v$(HOME)/.cfconfig:/root -v$(shell pwd):$(CODEFRESH_CLI_SHARED_DIR) $(CODEFRESH_CLI_IMAGE):$(CODEFRESH_CLI_VERSION)
```
Comment (osterman) [inline modules/codefresh/Makefile.sync]: Hrm...... maybe we should think about running this in the context of a codefresh pipeline. 

Can we do this in a GitOps fashion?

Comment (osterman) [reply modules/codefresh/Makefile.sync]: (a pipeline that creates pipelines)

File: templates/codefresh.sync.mask.yaml (hunk 6 above)
Comment (goruha) [reply templates/codefresh.sync.mask.yaml:3]: Codefresh specificiation

File: modules/codefresh/Makefile.sync (hunk 7 above)
Comment (osterman) [reply modules/codefresh/Makefile.sync]: e.g. can we support a codefresh pipeline that looks like this...

On PR:
```
commands:
  - make codefresh/sync/diff
```

on merge:
```
commands:
  - make codefresh/sync/apply
```

Comment (goruha) [reply modules/codefresh/Makefile.sync]: Great idea.

**Comment by osterman** [comment]
Please add this PR description to the `modules/codefresh/README.md`

File: bin/codefresh-sync.sh [hunk 8]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env bash
+
+## Messages templates
+MESSAGE_SEPARATOR="$(printf -- '_%.0s' {1..80})";
+
+MESSAGE_NO_CHANGES_REQUIRED=">>>>>>>> Processing ${PIPELINE} pipeline ... No changes required";
+
+MESSAGE_CHANGES_REQUIRED=">>>>>>>> Processing ${PIPELINE} pipeline ... Changes found
+${MESSAGE_SEPARATOR}
+This changes would be applied if you run the same command with environment variables APPLY=true
```
Comment (aknysh) [inline bin/codefresh-sync.sh]: These changes will be applied if you run the same command with the environment variable APPLY=true

**Review by aknysh — COMMENTED** [review]
see comments

File: bin/codefresh-sync.sh (hunk 8 above)
Comment (osterman) [inline bin/codefresh-sync.sh]:
```suggestion
These changes will be applied if you run the same command with the environment variable APPLY=true
```

File: bin/codefresh-sync.sh [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+or
+make codefresh/sync REPOSITORIES=${REPOSITORY} ACCOUNTS=${ACCOUNT} PIPELINES=${PIPELINE} APPLY=true
+${MESSAGE_SEPARATOR}";
+
+MESSAGE_APPLING_CHANGES=">>>>>>>> Processing ${PIPELINE} pipeline ... Changes applied
```
Comment (osterman) [inline bin/codefresh-sync.sh]:
```suggestion
MESSAGE_APPLING_CHANGES=">>>>>>>> Processing ${PIPELINE} pipeline... Changes applied
```

File: modules/codefresh/Makefile.sync [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+## Sync all
+codefresh/sync: codefresh/sync/deps
+	$(call assert-set,REPOSITORIES)
+	@for repo in $(REPOSITORIES) ; do \
+		echo ">>>> Processing $${repo} repo"; \
+		for acc in $(ACCOUNTS) ; do \
+			$(MAKE) -s codefresh/sync/apply ACCOUNT=$$acc REPOSITORY=$$repo; \
+		done; \
+		echo "============================================================================================================" ; \
```
Comment (osterman) [inline modules/codefresh/Makefile.sync]: Use `MESSAGE_SEPARATOR`

File: modules/codefresh/Makefile.sync [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+define codefresh-use-context
+	@$(CODEFRESH_CLI) auth use-context $(PROJECT)-$(1)
+endef
+
+## Authentificate on codefresh account
+codefresh/sync/auth/%:
+	$(call assert-set,CF_API_KEY)
+	$(call assert-set,PROJECT)
+	@$(CODEFRESH_CLI) auth create-context $(PROJECT)-$(subst codefresh/sync/auth/,,$@) --api-key $(CF_API_KEY)
```
Comment (osterman) [inline modules/codefresh/Makefile.sync:18]: I think if the ENV is `CF_API_KEY` we don't need to pass it on the command line

File: modules/codefresh/Makefile.sync [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+## Install dependencies for codefresh sync
+codefresh/sync/deps: packages/install/gomplate packages/install/yq
+	@exit 0
+
+## Sync all
+codefresh/sync: codefresh/sync/deps
+	$(call assert-set,REPOSITORIES)
+	@for repo in $(REPOSITORIES) ; do \
+		echo ">>>> Processing $${repo} repo"; \
+		for acc in $(ACCOUNTS) ; do \
```
Comment (osterman) [inline modules/codefresh/Makefile.sync]:
```suggestion
		for account in $(ACCOUNTS) ; do \
```

File: modules/codefresh/Makefile.sync [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+			$(MAKE) -s codefresh/sync/apply ACCOUNT=$$acc REPOSITORY=$$repo; \
```
Comment (osterman) [inline modules/codefresh/Makefile.sync]:
```suggestion
			$(MAKE) -s codefresh/sync/apply ACCOUNT=$$account REPOSITORY=$$repo; \
```

**Review by osterman — CHANGES_REQUESTED** [review]
Add a `README.md` to the `module/codefresh` repo with roughly these contents

```
## What
* Added targets to sync codefresh pipelines from code with templates

## Why
* To have a repeatable, stable  and scalable process of creation/update codefresh pipelines

## Usage
* Create new repo that would contain templates for your pipelines

~~~~
{repo}                                                           # Repository for codefresh sync
├── templates                                            # Directory for templates
│   ├── {account_x}                                    # Templates for account x
│   │    ├── {template_x_x1}.yaml              # Pipeline template {template_x_x1}
│   │    └── {template_x_y1}.yaml                # Pipeline template {template_x_y1}
│   └── {account_y}                                      # Templates for account y
│        ├── {template_y_x1}.yaml                # Pipeline template {template_y_x1}
│        └── {template_y_y1}.yaml                # Pipeline template {template_y_y1}
└── Makefile                                               # Makefile
```
* Fill templates with [pipelines](https://codefresh-io.github.io/cli/pipelines/spec/). You can use `{{ (ds "repository") }}` as placeholder for repo name
* Add to the `Makefile` 
```
export DEFAULT_HELP_TARGET ?= help
export HELP_FILTER ?= help|init|codefresh/sync
export DOCKER_NETWORK ?= default
export BUILD_HARNESS_BRANCH ?= 0.14.5
-include $(shell curl -sSL -o .build-harness "https://git.io/build-harness"; echo .build-harness)

## Project name
PROJECT ?= project_z

## Repos to sync pipelines
REPOSITORIES ?= api \
				front
## Codefresh accounts to operate with
ACCOUNTS ?= account_x account_y

## Pipelines for `acccount_x`
ACCOUNT_X_PIPELINES ?= template_x_x1 template_x_y1

## Pipelines for `acccount_y`
ACCOUNT_Y_PIPELINES ?= template_y_x1 template_y_y1
```
* Do Auth for accounts. You can get `CF_API_KEY` [here](https://g.codefresh.io/account-conf/tokens)
```
make codefresh/sync/auth/account_x CF_API_KEY={CF_API_KEY for `account_x`}
make codefresh/sync/auth/account_y CF_API_KEY={CF_API_KEY for `account_y`}
```

* For dry-run use
```
make codefresh/sync
```

* For sync use 
```
make codefresh/sync APPLY=true
```

* You can filter repos, accounts and pipelines that are syncing with env variables `REPOSISTORIES` `ACCOUNTS` `PIPELINES`

```
make codefresh/sync REPOSITORIES=api ACCOUNTS=account_x PIPELINES=template_x_x1
```

~~~~

File: modules/chamber/Makefile [hunk 14]
```diff
@@ -11,7 +11,8 @@ endif
 WITH_CHAMBER = $(CHAMBER_AUTH) chamber exec $(CHAMBER_SERVICES) --
 
 ## Install chamber
-chamber/install: packages/install/chamber
+chamber/install:
+	@$(MAKE) -s packages/install/chamber
```
Comment (osterman) [inline modules/chamber/Makefile]: Why are you changing syntax here?

File: modules/gitleaks/Makefile [hunk 15]
```diff
@@ -3,7 +3,8 @@ export GITLEAKS_VERSION ?= 1.2.0
 GITLEAKS ?= gitleaks
 
 ## Install gitleaks
-gitleaks/install: packages/install/gitleaks
+gitleaks/install:
+	@$(MAKE) -s packages/install/gitleaks
```
Comment (osterman) [inline modules/gitleaks/Makefile]: Why changing the syntax?

File: modules/docs/Makefile [hunk 16]
```diff
@@ -14,5 +14,6 @@ docs/targets.md: docs/deps
 
 .PHONY : docs/terraform.md
 ## Update `docs/terraform.md` from `terraform-docs`
-docs/terraform.md: docs/deps packages/install/terraform-docs
+docs/terraform.md: docs/deps
+	@$(MAKE) -s packages/install/terraform-docs
```
Comment (osterman) [inline modules/docs/Makefile]: Why are you changing syntax?

**Review by osterman — CHANGES_REQUESTED** [review]
See comments. Also, many unaddressed previous comments.

**Comment by osterman** [comment]
- [x] add readme
- [x] resolve comments 
- [x] restore dep deps

File: modules/chamber/Makefile (hunk 14 above)
Comment (goruha) [reply modules/chamber/Makefile]: It was debug issue.
Reverted.

File: modules/codefresh/Makefile.sync (hunk 11 above)
Comment (goruha) [reply modules/codefresh/Makefile.sync:18]: No. `CF_API_KEY` env overrides key for the current context, but to do `auth` need to specify it implicitly.

File: modules/codefresh/README.md [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+* Added targets to sync codefresh pipelines from code with templates
+
+## Why
+* To have a repeatable, stable  and scalable process of creation/update codefresh pipelines
+
+## Usage
+* Create new repo that would contain templates for your pipelines
+
+```
+{repo}                                                           # Repository for codefresh sync
```
Comment (osterman) [inline modules/codefresh/README.md]: Fix spacing

File: bin/codefresh-sync.sh [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+	touch ${PIPELINE_CURRENT}
+	## New pipeline can be applied without any changes
+	cp  ${PIPELINE_NEW} ${PIPELINE_TO_APPLY}
+else
+
+	## Get current pipeline
+	${CODEFRESH_CLI} get pipelines ${PIPELINE_FULLNAME} -o yaml > ${PIPELINE_CURRENT}
+
+	## Create a copy of pipelines to apply
+  cp ${PIPELINE_NEW} ${PIPELINE_TO_APPLY}
```
Comment (osterman) [inline bin/codefresh-sync.sh:52]: Fix whitespace

File: bin/codefresh-sync.sh [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+MESSAGE_NO_CHANGES_REQUIRED=">>>>>>>> Processing ${PIPELINE} pipeline ... No changes required";
+
+MESSAGE_CHANGES_REQUIRED=">>>>>>>> Processing ${PIPELINE} pipeline ... Changes found
+${MESSAGE_SEPARATOR}
+These changes will be applied if you run the same command with the environment variable APPLY=true
+or
+make codefresh/sync/apply REPOSITORIES=${REPOSITORY} ACCOUNTS=${ACCOUNT} PIPELINES=${PIPELINE}
+${MESSAGE_SEPARATOR}";
+
+MESSAGE_APPLING_CHANGES=">>>>>>>> Processing ${PIPELINE} pipeline... Changes applied
```
Comment (osterman) [inline bin/codefresh-sync.sh:15]: Use `\n` here and instead `echo -e $VAR` which will then interpret the `\n`

**Review by osterman — CHANGES_REQUESTED** [review]
See inline comments

File: bin/codefresh-sync.sh (hunk 19 above)
Comment (goruha) [reply bin/codefresh-sync.sh:15]: I tried. 
It looks ugly - too long.
It's easy to see what you get with current solution

**Review by osterman — CHANGES_REQUESTED** [review]
using the combination of tabs & spaces. let's move to tabs since that's what the `shfmt` will look for (in the future when we implement it like in geodesic). https://github.com/cloudposse/geodesic/blob/master/.travis.yml#L19

File: modules/codefresh/README.md [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+```
+{repo}                                  # Repository for codefresh sync
+├── templates                           # Directory for templates
+│   ├── {account_x}                     # Templates for account x
+│   │    ├── {template_x_x1}.yaml       # Pipeline template {template_x_x1}
+│   │    └── {template_x_y1}.yaml       # Pipeline template {template_x_y1}
+│   └── {account_y}                     # Templates for account y
+│        ├── {template_y_x1}.yaml       # Pipeline template {template_y_x1}
+│        └── {template_y_y1}.yaml       # Pipeline template {template_y_y1}
+└── Makefile                             # Makefile
```
Comment (osterman) [inline modules/codefresh/README.md]: Space

File: modules/codefresh/README.md [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+* Added targets to sync codefresh pipelines from code with templates
+
+## Why
+* To have a repeatable, stable  and scalable process of creation/update codefresh pipelines
+
+## Usage
+* Create new repo that would contain templates for your pipelines
+
+```
+{repo}                                  # Repository for codefresh sync
```
Comment (osterman) [inline modules/codefresh/README.md:11]: space

Comment (goruha) [reply modules/codefresh/README.md:11]: I do not the problem

**Review by osterman — APPROVED** [review]

**Pull request merged**